libs/platform/ebpf_epoch.c

```diff
@@ -323,95 +332,73 @@ ebpf_epoch_terminate()
-    ebpf_assert(is_preemptible == ebpf_is_preemptible());
-    ebpf_lock_state_t state = ebpf_lock_lock(&_ebpf_epoch_cpu_table[current_cpu].lock);
+    // Assert that the BPF program didn't change the IRQL.
+    ebpf_assert(KeGetCurrentIrql() == epoch_state->irql_at_enter);
 
-    // Mark the epoch state as available.
-    ebpf_assert(epoch_state->epoch != 0);
-    epoch_state->epoch = 0;
-    epoch_state->is_preemptible = false;
+    if (old_irql < DISPATCH_LEVEL) {
+        old_irql = KeRaiseIrqlToDpcLevel();
+    }
 
-    // Mark the epoch state as not stale.
-    _ebpf_epoch_cpu_table[current_cpu].stale = false;
+    uint32_t cpu_id = old_irql < DISPATCH_LEVEL ? ebpf_get_current_cpu() : epoch_state->cpu_id;
 
-    if (!ebpf_list_is_empty(&_ebpf_epoch_cpu_table[current_cpu].free_list)) {
-        release_free_list = true;
-    }
+    // Special case: Thread has moved to a different CPU since entering the epoch.
+    if (cpu_id != epoch_state->cpu_id) {
+        // Signal the other CPU to remove the thread entry.
+        ebpf_assert(epoch_state->irql_at_enter < DISPATCH_LEVEL);
+        if (old_irql < DISPATCH_LEVEL) {
+            KeLowerIrql(old_irql);
+        }
 
-    ebpf_lock_unlock(&_ebpf_epoch_cpu_table[current_cpu].lock, state);
-    if (release_free_list) {
-        _ebpf_epoch_release_free_list(&_ebpf_epoch_cpu_table[current_cpu], _ebpf_release_epoch);
-    }
+        ebpf_epoch_cpu_message_t message = {0};
+        message.message_type = EBPF_EPOCH_CPU_MESSAGE_TYPE_EXIT_EPOCH;
+        message.message.exit_epoch.epoch_state = epoch_state;
 
-    if (is_preemptible) {
-        // Release the thread entry.
-        ebpf_semaphore_release(_ebpf_epoch_cpu_table[current_cpu].epoch_table_semaphore);
+        // The other CPU will call ebpf_epoch_exit() on our behalf.
+        // This will trigger the assert at the top of the function because
+        // the IRQL will be DISPATCH_LEVEL. Set the irql_at_enter to DISPATCH_LEVEL
+        // to avoid the assert.
+        KIRQL saved_irql = epoch_state->irql_at_enter;
+        epoch_state->irql_at_enter = DISPATCH_LEVEL;
 
-        // Allow APCs to run.
-        ebpf_leave_critical_region();
-    }
-}
+        _ebpf_epoch_send_message_and_wait(&message, epoch_state->cpu_id);
 
-void
-ebpf_epoch_flush()
-{
-    int64_t released_epoch;
-    if (!_ebpf_epoch_cpu_table) {
+        // Restore the irql at enter.
+        epoch_state->irql_at_enter = saved_irql;
         return;
     }
 
-    _ebpf_epoch_get_release_epoch(&released_epoch);
-    EBPF_LOG_MESSAGE_UINT64(
-        EBPF_TRACELOG_LEVEL_VERBOSE, EBPF_TRACELOG_KEYWORD_EPOCH, "_ebpf_release_epoch updated", released_epoch);
-    // _ebpf_release_epoch is updated outside of any lock.
-    _ebpf_release_epoch = released_epoch;
+    ebpf_list_remove_entry(&epoch_state->epoch_list_entry);
+    _ebpf_epoch_release_free_list(&_ebpf_epoch_cpu_table[cpu_id], _ebpf_release_epoch);
+
+    _ebpf_epoch_lower_to_previous_irql(epoch_state->irql_at_enter);
```

Comment:
why does the irql_at_enter needs to be reset to previous irql? In fact shouldnt the epoch_state be scrubbed?

Reply:
Added more clarification.